### PR TITLE
Add optional `importer` parameter to `this.emitFile`

### LINKED ISSUE #3436: Should emitFile take an importer?
### Feature Use Case

Take plugin code like this:

```js
const prefix = 'typescript:';
{
  async resolveId(id, importer) {
    if (!id.startsWith(prefix)) return null;

    const realId = id.slice(prefix.length);
    const resolveResult = await this.resolve(realId, importer);
    if (!resolveResult) {
      throw Error(`Cannot resolve ${realId}`);
    }
    return prefix + resolveResult.id;
  },
}
```

Which converts `typescript:./whatever.ts` to `typescript:/full/path/to/whatever.ts`, based on the importer.

However, another plugin cannot invoke this via `emitFile`:

```js
this.emitFile({
  type: 'chunk',
  id: 'typescript:./whatever.ts',
});
```

The other plugin would have to understand the `typescript:` prefix, and handle resolving itself.

### Feature Proposal

```js
this.emitFile({
  type: 'chunk',
  id: 'typescript:./whatever.ts',
  importer: '…',
});
```

The value of `importer` would be passed to `resolveId`.

### PR DESCRIPTION
**Pull request by lukastaegert**
This PR contains:
- [ ] bugfix
- [x] feature
- [ ] refactor
- [x] documentation
- [ ] other

Are tests included?
- [x] yes (*bugfixes and features will not be merged without tests*)
- [ ] no

Breaking Changes?
- [ ] yes (*breaking changes will not be merged unless absolutely necessary*)
- [x] no

List any relevant issue numbers:
Resolves #3436 

### Description
This adds the possibility to specify an importing file when using `this.emitFile` so that relative ids can be resolved correctly. Also cleans up documentation and types to specify that `importer` can actually be `undefined`, in which case the current working directory is used as reference.

### CODE REVIEW

**Comment by jakearchibald** [comment]
I haven't tested it, but at a glance this is exactly what I was asking for!

File: docs/05-plugin-development.md [hunk 1]
```diff
@@ -468,7 +469,9 @@ You can reference the URL of an emitted file in any code returned by a [`load`](
...
 
-If the `type` is *`chunk`*, then this emits a new chunk with the given module id as entry point. This will not result in duplicate modules in the graph, instead if necessary, existing chunks will be split or a facade chunk with reexports will be created. Chunks with a specified `fileName` will always generate separate chunks while other emitted chunks may be deduplicated with existing chunks even if the `name` does not match. If such a chunk is not deduplicated, the [`output.chunkFileNames`](guide/en/#outputchunkfilenames) name pattern will be used.
+If the `type` is *`chunk`*, then this emits a new chunk with the given module `id` as entry point. To resolve it, the `id` will be passed through all [`resolveId`](guide/en/#resolveid) hooks, just like regular entry points. If an `importer` is provided, this acts as the second parameter of `resolveId` and is important to properly resolve relative paths. If it is not provided, paths will be resolved relative to the current working directory.
```
Comment (jakearchibald) [inline docs/05-plugin-development.md]:
```suggestion
If the `type` is *`chunk`*, then this emits a new chunk with the given module `id` as entry point. To resolve it, the `id` will be passed through build hooks just like regular entry points, starting with [`resolveId`](guide/en/#resolveid). If an `importer` is provided, this acts as the second parameter of `resolveId` and is important to properly resolve relative paths. If it is not provided, paths will be resolved relative to the current working directory.
```

Dunno if this is better/worse. Up to you.

Comment (lukastaegert) [reply docs/05-plugin-development.md]: 👍

**Comment by jakearchibald** [comment]
Thank you for this!

**Pull request merged**